crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -667,6 +667,50 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
         Ok(deleted)
     }
 
+    /// Prune the dupsort table for the specified pre-sorted key iterator, calling `chunk_callback`
+    /// after every `batch_size` pruned rows (subkeys).
+    ///
+    /// Returns number of rows (subkeys) pruned.
+    pub fn prune_dupsort_table_in_batches<T, K>(
+        &self,
+        keys: impl IntoIterator<Item = K>,
+        batch_size: usize,
+        mut batch_callback: impl FnMut(usize),
+    ) -> std::result::Result<usize, DatabaseError>
+    where
+        T: DupSort<Key = K>,
+        K: Key,
+    {
+        let mut cursor = self.tx.cursor_dup_write::<T>()?;
+        let mut deleted = 0;
+
+        for key in keys {
+            if cursor.seek(key)?.is_some() {
```

Comment:
although its current usage seems fine - `seek` gets `greater than or equal to key.` and we're not comparing it

Reply:
actually, we might be deleting too much. if a block has no changesets, this will delete the elements of the next one

Reply:
`seek_exact` seems to be working fine, it points to the first subkey of the dupsort table for the provided key